### ISSUE TITLE
IContextProvider and GenContext improvements

### ISSUE THREAD

**Issue by sibille** [Microsoft/WindowsTemplateStudio#2396]
The different paths in IContextProvider is hard to understand, so we should: 

- [x] Move all generation logic to core 
- [x] Rename outputPath to GenerationOutputPath
- [x] Remove TempGenerationPath and DestinationParentPath and resolve them dynamically
- [x] Add checks to verify paths are set correctly before starting generation
- [x] Add some documentation

**Comment by mrlacey** [comment]
There should only be a need to set the directory the solution is created in and everything else should be inferred/generated.

**Comment by mrlacey** [comment]
Issues with the current implementation were raised as part of the parallel work being done for v3.0 
Can the changes proposed here please be deferred so they don't create a merge headache with the 3.0 work

**Comment by sibille** [comment]
Sure. Another option is to do this work on the 3.0 branch.

**Comment by sibille** [comment]
We had a look at the paths, and can infer other paths from DestinationPath (pointing to the project) and GenerationOutputPath (pointing to the project for new projects and pointing to the temp folder for new items).

**Comment by sibille** [comment]
Added documentation on core project and GenContext/IContextProvider to https://github.com/Microsoft/WindowsTemplateStudio/blob/dev/docs/getting-started-developers.md
